jbmc/src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -2335,105 +2335,78 @@ codet &java_bytecode_convert_methodt::do_exception_handling(
+        // generating a label corresponding to the handler's pc
+        label(std::to_string(exception.handler_pc)));
+    }
+    // Prepend the CATCH-PUSH instruction
+    code = code_blockt({ catch_push, code });
   }
 
-  // next add the CATCH-POP instructions
-  size_t start_pc = 0;
-  // as the first try block may have start_pc 0, we
-  // must track it separately
-  bool first_handler = false;
-  // check if this is the end of a try block
+  // Next add the CATCH-POP instructions
+  // exception_row.end_pc is exclusive so append CATCH-POP instructions if this
+  // is the instruction before it.
+  // Dangerously, we assume that the lexically next bit of code after the try
+  // block (the catch block) is:
+  // a) the next code in the bytecode and therefore its start address matches
+  // the exclusive end position of the try block (probably a valid assumption
+  // for non-obfuscated bytecode generated by most compilers)
+  // b) the successor of a previous instruction and therefore in the working
+  // set (not a valid assumption when there aren't any throwing statements in
+  // the try block)
+  // We are therefore going to find all exception handlers that end at our
+  // earliest successor
+  if(working_set.empty())
+  {
+    // There was no throw in the try block
+    // TODO: This will result in non-balanced PUSH and POPs
+    return code;
+  }
+  const method_offsett &working_set_elt = *working_set.begin();
+  if(cur_pc >= working_set_elt)
+  {
+    // This just fails if we have a back-edge here
+    // TODO: This will result in non-balanced PUSH and POPs
+    return code;
+  }
+  // Count the distinct handler start positions that end at this location
+  std::set<std::size_t> start_positions;
   for(const auto &exception_row : method.exception_table)
   {
-    // add the CATCH-POP before the end of the try block
-    if(
-      cur_pc < exception_row.end_pc && !working_set.empty() &&
-      *working_set.begin() == exception_row.end_pc)
-    {
-      // have we already added a CATCH-POP for the current try-catch?
-      // (each row corresponds to a handler)
-      if(exception_row.start_pc != start_pc || !first_handler)
-      {
-        if(!first_handler)
-          first_handler = true;
-        // remember the beginning of the try-catch so that we don't add
-        // another CATCH-POP for the same try-catch
-        start_pc = exception_row.start_pc;
-        // add CATCH_POP instruction
-        code_pop_catcht catch_pop;
-        c = code_blockt({c, catch_pop});
-      }
-    }
+    // Check if working_set_elt is the end of a try block
+    if(working_set_elt == exception_row.end_pc)
+      start_positions.insert(exception_row.start_pc);
   }
-  return c;
+  for(std::size_t handler = 0; handler < start_positions.size(); ++handler)
+    // Append a CATCH-POP instruction before the end of the try block
+    code = code_blockt({ code, code_pop_catcht() });
```

Comment:
:bulb: That's a lot of nested blocks, would it make sense to produce `code; pop; pop; pop` rather than `{{{code; pop} pop} pop}`?

E.g. by:

```C++
auto pop_catchs = std::vector<codet>{start_positions.size(), code_pop_catcht()};
code = code_blockt{code, pop_catchs)
```

Reply:
Managed to do this for both pushes and pops, thank you.

Reply:
Unfortunately this was actually wrong. Reverted and gone with something like your initial idea.